Utility: use Fsdk

webwarrior-ws opened the pull request:
Added fsdk package. Reuse functions from Fsdk.

Renamed FSharpUtil module to AsyncUtil.

Changed Retry function to take 1 type parameter instead of 2 as handling 2 types of exceptions is not needed anymore because SocketException is wrapped in NOnionException.

webwarrior-ws committed:
Utility: made Retry func take 1 type parameter

Changed Retry function to take 1 type parameter instead of 2 as
handling 2 types of exceptions is not needed anymore because
SocketException is wrapped in NOnionException.

aarani commented:
LGTM

knocte commented on NOnion/Utility/FSharpUtil.fs:
```diff
@@ -65,7 +65,7 @@ module FSharpUtil =
         }
 
-    let Retry<'E1, 'E2 when 'E1 :> Exception and 'E2 :> Exception>
+    let Retry<'E1 when 'E1 :> Exception>
```
@webwarrior-ws if there's only 1 exception type parameter now, there's no point in naming it E1, please rename it to TEx

knocte commented:
@webwarrior-ws can you also fix the conventions job with a similar commit to this please: [37e374a7b03cb55b4718c46e65911f09c21f876b](https://github.com/nblockchain/geewallet/commit/37e374a7b03cb55b4718c46e65911f09c21f876b) (and move it to be the first commit).

webwarrior-ws commented:
Should  be good now

knocte commented:
> LGTM

Please approve via GitHub PR approval process.

webwarrior-ws committed:
Network,Services,Utility: use Fsdk

Added fsdk package. Reuse functions from Fsdk.

webwarrior-ws commented:
Can't use Fsdk because it requires FSharp.Core 6.0, but NOnion uses 4.7.2 (in CI at least).

knocte commented:
Upgrade FSharp.Core then?

knocte commented on NOnion/Utility/FSharpUtil.fs:
```diff
@@ -4,94 +4,41 @@ open System
-            | None ->
-                // none of the jobs passed to Async.Choice returns None
-                return failwith "unreachable"
+            match result with
```
@webwarrior-ws why keep WithTimeout here instead of directly use the one in Fsdk?

webwarrior-ws replied:
Because of different semantics. Fsdk variant returns `option`, but NOnion's variant returns value or raises `TimeoutErrorException`.

knocte replied:
well, the idea of the initial request about combining both NOnion and geewallet's FSharpUtil modules into just one means that we should keep the best approach in each case; so let's decide what's the best approach and have only one (if it requires changing all callers of the func in geewallet or NOnion, that's fine)

webwarrior-ws replied:
Changed NOnion to use Fsdk variant.

knocte commented:
Thinking about it more, please revert your last commit that removes NOnion.Utility.FSharpUtil.WithTimeout, but I want FSharpUtil module be gone from NOnion to prevent naming clash with Fsdk, so let's rename it to NetworkUtils.fs? (To be consistent with other files in that folder.) BTW I had a look at other files in that folder and I think ResultUtils.fs's OperationResult overlaps with FSharpUtil's Either, let's remove it and use Either. And same thing for the Unwrap function, I think FSharpUtil has Unwrap func? If not, let's move it to Fsdk.

webwarrior-ws committed:
Http,Network,Services.Utitlity: rename FSharpUtil

Renamed FSharpUtil module to AsyncUtil.

aarani commented on NOnion/Utility/NetworkUtils.fs:
```diff
@@ -1,14 +1,12 @@
 open NOnion
 
-module FSharpUtil =
+module NetworkUtils =
```
Let's call this AsyncUtil instead, WithTimeout is not directly related to Network.

webwarrior-ws replied:
Renamed to `AsyncUtil`.

webwarrior-ws committed:
Utility,Directory: use resources func from Fsdk

Use ExtractEmbeddedResourceFileContents function from Fsdk.
Remove local copy of that function.

knocte commented:
> Thinking about it more, please revert your last commit that removes NOnion.Utility.FSharpUtil.WithTimeout, but I want FSharpUtil module be gone from NOnion to prevent naming clash with Fsdk, so let's rename it to NetworkUtils.fs? (To be consistent with other files in that folder.) BTW I had a look at other files in that folder and I think ResultUtils.fs's OperationResult overlaps with FSharpUtil's Either, let's remove it and use Either. And same thing for the Unwrap function, I think FSharpUtil has Unwrap func? If not, let's move it to Fsdk.

Was this done?

webwarrior-ws commented:
> > Thinking about it more, please revert your last commit that removes NOnion.Utility.FSharpUtil.WithTimeout, but I want FSharpUtil module be gone from NOnion to prevent naming clash with Fsdk, so let's rename it to NetworkUtils.fs? (To be consistent with other files in that folder.) BTW I had a look at other files in that folder and I think ResultUtils.fs's OperationResult overlaps with FSharpUtil's Either, let's remove it and use Either. And same thing for the Unwrap function, I think FSharpUtil has Unwrap func? If not, let's move it to Fsdk.
> 
> Was this done?

First part - yes, FSharpUtil is now called AsyncUtil.
OperationsResult is still there; it has different signature (1 type parameter instead of 2 in Fsdk.Either).
